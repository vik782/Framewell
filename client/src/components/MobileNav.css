:root {
    --mobile-nav-height: 300px;
}

#mobilenav {
    background: #96b3c2;
    padding: 20px;
    height: var(--mobile-nav-height);
    transition: 1s;
    right: 0;
    box-sizing: border-box;
    align-items: center;
    justify-content: center;
    z-index: 10;
    box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.2);
    display: none;
    transform-origin: top center;
    animation: growDown 1200ms ease-in-out forwards;
    overflow-x: hidden;
}

.mobilenav-exit {
    display: flex;
    margin: 30px;
    height: 50px;
    align-items: center;
    cursor: pointer;
}

#exitNavIcon {
    color: white;
    font-size: 25px;
}

.mobilenav-content {
    display: flex;
    align-items: center;
    justify-content: space-between;
    flex-direction: column;
    height: calc(var(--mobile-nav-height) - 160px);
}

.mobilenav-link {
    cursor: pointer;
    text-decoration: none;
    color: #ffffff;
    position: relative;
    font-weight: 700;
}

.mobilenav-content .mobilenav-link > i {
    margin-right: 10px;
    font-size: 18px;
    color: #ffffff;
}

.mobilenav-link::before {
    content: '';
    position: absolute;
    bottom: 0;
    right: 0;
    width: 0;
    height: 1px;
    background-color: white;
    transition: width 0.6s cubic-bezier(0.25, 1, 0.5, 1);
}

.mobilenav-responsive {
    display: inline !important;
    z-index: 9999 !important;
    left: 0 !important;
    position: absolute;
}

/* Animations */
@keyframes growDown {
    0% {
        height: 0;
        margin: -14px;
        padding: 0;
    }
    100% {
        height: var(--mobile-nav-height);
        margin: 0;
        padding: 0;
    }
}

@media (hover: hover) and (pointer: fine) {
    .mobilenav-link:hover::before{
      left: 0;
      right: auto;
      width: 100%;
      color: white;
    }
  }
