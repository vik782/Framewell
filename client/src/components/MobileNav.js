/**
 * @fileoverview Abstraction of the Navbar component; the component that the user sees
 *               when they view the page in their mobile device
 * Uses:
 * - React for rendering HTML
 * - Iconify for adding icons
 * - React Router for handling client-side routes
 * - Universal Cookie for handling browser cookies and validating logins
 */

// Imports of packages
import React from "react";
import { Icon } from "@iconify/react";
import { Link } from "react-router-dom";
import Cookies from "universal-cookie";

// Style-based imports
import "./MobileNav.css";

// obtain token from cookie
const cookies = new Cookies();

/**
 * The mobile nav takes in props that decide whether it should show or not
 * and contains links that the user can use to navigate accordingly
 * @return {React.Component}
 */
function MobileNav({ mobileNavOpen, closeMobileNav }) {
  /**
   * Removes the cookie that authenticated the login and redirects the user
   * back to the login page
   */
  function logout() {
    cookies.remove("TOKEN", { path: "/" });

    // user redirected to login page
    window.location.href = "/";
  }

  let /** React.Component */ DashboardButton = window.location.href.includes(
      "/dashboard"
    ) ? (
      <span
        className="mobilenav-link"
        onClick={() => window.location.reload()}
      >
        Dashboard
      </span>
    ) : (
      /* else */ 
      <Link className="mobilenav-link" to={`/dashboard`}>
        Dashboard
      </Link>
    );

  return (
    <div className={mobileNavOpen ? "mobilenav-responsive" : ""}
      id="mobilenav"
    >
      <div className="mobilenav-exit">
        <Icon
          icon="akar-icons:cross"
          id="exitNavIcon"
          onClick={() => closeMobileNav()}
        />
      </div>

      <div className="mobilenav-content">
        <Link className="mobilenav-link" to={`/dashboard`}>
          Dashboard
        </Link>
        <Link className="mobilenav-link" to={`/add-artefact`}>
          Add Artefact
        </Link>
        <div className="mobilenav-link">
          <i>
            <Icon icon="icon-park-outline:logout" />
          </i>
          <span onClick={() => logout()}>Sign Out</span>
        </div>
      </div>
    </div>
  );
}

export default MobileNav;
