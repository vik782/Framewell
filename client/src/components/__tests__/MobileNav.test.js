import React from "react";
import MobileNav from "../MobileNav";
import { render, fireEvent, getByTestId } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

const resizeWindow = (x) => {
    window.innerWidth = x;
    window.dispatchEvent(new Event('resize'));
};

window.getComputedStyle = window.getComputedStyle || function( element ) {
    return element.currentStyle;
}

// Snapshot test to account for changes to the UI, whether expected or unexpected
test("MobileNav renders correctly", () => {
    const mobileNav= renderer.create(<MemoryRouter><MobileNav /></MemoryRouter>);
    expect(mobileNav).toMatchSnapshot();
});

// Unit Testing
test("MobileNav is rendered only when media query is below 978px", () => {
    const { getByTestId } = render(<MemoryRouter><MobileNav /></MemoryRouter>);
    const mobileNav = getByTestId("mobile-nav");

    // Resize window to media query 978px
    resizeWindow(600);
    expect(mobileNav).toHaveStlye("display: inline");

    // Resize window to desktop view
    resizeWindow(600);
    expect(mobileNav).toHaveStlye("display: none");
});

test("Exit MobileNav button exits correctly", () => {
    const closeMobileNavSpy = jest.fn();
    const { getByTestId } = render(<MemoryRouter><MobileNav closeMobileNav={closeMobileNavSpy}/></MemoryRouter>);
    const exitMobile = getByTestId("mobile-exit");

    fireEvent.click(exitMobile);
    expect(closeMobileNavSpy).toHaveBeenCalled();
});