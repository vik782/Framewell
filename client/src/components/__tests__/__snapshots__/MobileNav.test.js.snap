// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileNav renders correctly 1`] = `
<div
  className="mobilenav"
  data-testid="mobile-nav"
>
  <div
    className="mobilenav-exit"
    data-testid="mobile-exit"
    onClick={[Function]}
  >
    <span />
  </div>
  <div
    className="mobilenav-content"
  >
    <a
      className="mobilenav-link"
      href="/dashboard"
      onClick={[Function]}
    >
      <a>
        Dashboard
      </a>
    </a>
    <a
      className="mobilenav-link"
      href="/add-artefact"
      onClick={[Function]}
    >
      <a>
        Add Artefact
      </a>
    </a>
    <div
      className="mobilenav-link"
    >
      <i>
        <span />
      </i>
      <a
        onClick={[Function]}
      >
        Sign Out
      </a>
    </div>
  </div>
</div>
`;
